---
title: "Star Allele Comparison"
output: html_document
date: "2023-05-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries

```{r}
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ursaPGx))
```

## Generate calls for CYP2C8, CYP2C9, and CYP2C19 with `ursaPGx`

```{r}
# Specify the path to the VCF files
vcf <- "../data-raw/sample_vcf/1kGP_high_coverage_Illumina.chr10.filtered.SNV_INDEL_SV_phased_panel.vcf.gz"

# Genes to call star alleles for
genes <- c("CYP2C8", "CYP2C9", "CYP2C19")

# Create a single DataFrame from the results of the caller pipeline
result <- do.call(cbind, lapply(genes, function(x) callDiplotypes(vcf, gene = x)))

# Convert the result to a data.table
result <- as.data.table(result, keep.rownames = "Coriell")

# Rename the results with distinct colnames 
setnames(
  result, 
  old = c("CYP2C8", "CYP2C9", "CYP2C19"), 
  new = c("CYP2C8_ursapgx", "CYP2C9_ursapgx", "CYP2C19_ursapgx")
  )
```

## Read in the GeT-RM comparison paper data

[CYP2C8, CYP2C9, and CYP2C19 Characterization Using Next-Generation Sequencing and Haplotype Analysis](https://pubmed.ncbi.nlm.nih.gov/35134542/) is used as the comparison 
set for this script. The Supplementary data 3 from that paper includes separate 
sheets for the callsets for CYP2C8, CYP2C9, and CYP2C19.

```{r}
# Download the supplementary data from the paper
url <- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9069873/bin/mmc3.xlsx"
suppfile <- "mmc3.xlsx"
if (!file.exists(suppfile)) download.file(url, suppfile)

# Read in the data for each gene
getrm_CYP2C8 <- setDT(readxl::read_excel(suppfile, sheet = "CYP2C8", skip = 3))
getrm_CYP2C9 <- setDT(readxl::read_excel(suppfile, sheet = "CYP2C9", skip = 3))
getrm_CYP2C19 <- setDT(readxl::read_excel(suppfile, sheet = "CYP2C19", skip = 3)) 

# Select only the necessary columns from the supp data
getrm_CYP2C8 <- getrm_CYP2C8[, c(1, 3, 4)]
getrm_CYP2C9 <- getrm_CYP2C9[, c(1, 3, 5)]
getrm_CYP2C19 <- getrm_CYP2C19[, c(1, 3, 5)]
names(getrm_CYP2C8) <-  c("Coriell", "CYP2C8_getrm", "CYP2C8_ngs")
names(getrm_CYP2C9) <-  c("Coriell", "CYP2C9_getrm", "CYP2C9_ngs")
names(getrm_CYP2C19) <- c("Coriell", "CYP2C19_getrm", "CYP2C19_ngs")
```

## Merge all data into a single table

```{r}
# Join all data into single data.table
supp_df <- getrm_CYP2C8[getrm_CYP2C9, on = "Coriell"][getrm_CYP2C19, on = "Coriell"]
comparison_df <- result[supp_df, on = "Coriell"]

# Reorder the columns in the table to make it easier to compare
cols <- c("Coriell",
          "CYP2C8_getrm", "CYP2C8_ngs", "CYP2C8_ursapgx", 
          "CYP2C9_getrm", "CYP2C9_ngs", "CYP2C9_ursapgx", 
          "CYP2C19_getrm", "CYP2C19_ngs", "CYP2C19_ursapgx")
setcolorder(comparison_df, cols)
```

## Write comparison table out to a file

```{r}
fwrite(comparison_df, "star-allele-comparison.tsv", sep = "\t")
```
